source/openpulse/openpulse/parser.py

```diff
@@ -0,0 +1,264 @@
+            ]
+            if ctx.hardwareQubitList()
+            else []
+        )
+        return_type = (
+            self.visit(ctx.returnSignature().scalarType()) if ctx.returnSignature() else None
+        )
+
+        return ast.CalibrationDefinition(
+            name=_visit_identifier(ctx.Identifier()),
+            arguments=arguments,
+            qubits=qubits,
+            return_type=return_type,
+            body=statements,
+        )
+
+    def visitIndexOperator(self, ctx: openpulseParser.IndexOperatorContext):
+        if ctx.setExpression():
+            return self.visit(ctx.setExpression())
+
+        index_element = (
+            openpulseParser.ExpressionContext,
+            openpulseParser.RangeExpressionContext,
+        )
+
+        def predicate(child):
+            return isinstance(child, index_element)
+
+        return [self.visit(child) for child in ctx.getChildren(predicate=predicate)]
+
+    @span
+    def visitRangeExpression(self, ctx: openpulseParser.RangeExpressionContext):
+        # start, end and step are all optional as in [:]
+        # It could be [start:end] or [start:step:end]
+        start = None
+        end = None
+        step = None
+        colons_seen = 0
+
+        for child in ctx.getChildren():
+            if isinstance(child, openpulseParser.ExpressionContext):
+                expression = self.visit(child)
+                if colons_seen == 0:
+                    start = expression
+                elif colons_seen == 1:
+                    end = expression
+                else:
+                    step = end
+                    end = expression
+            elif child.getText() == ":":
+                colons_seen += 1
+
+        return ast.RangeDefinition(start=start, end=end, step=step)
+
+    @span
+    def visitReturnStatement(self, ctx: qasm3Parser.ReturnStatementContext):
+        if not self._in_subroutine() and not self._in_defcal():
+            _raise_from_context(ctx, "'return' statement outside subroutine or defcal")
+        if ctx.expression():
+            expression = self.visit(ctx.expression())
+        elif ctx.measureExpression():
+            expression = self.visit(ctx.measureExpression())
+        else:
+            expression = None
+        return ast.ReturnStatement(expression=expression)
+
+
+# Try to reuse some methods from QASMNodeVisitor
+# We exclude some methods. The first 3 are from the generated openpulseParserVisitor class.
+# The rest are redefined in the OpenPulseNodeVisitor class.
+# We need to redefine some methods does instanceof checks.
+p = re.compile("_?(visit).+")
+excluded = [
+    "visitChildren",
+    "visitErrorNode",
+    "visitTerminal",
+    "visitArrayLiteral",
+    "visitCalibrationGrammarStatement",
+    "visitDefcalStatement",
+    "visitIndexOperator",
+    "visitRangeExpression",
+    "visitReturnStatement",
+    "visitScalarType",
+]
+for m in QASMNodeVisitor.__dict__:
```

Comment:
We import most methods from `openqasm3.parser.QASMNodeVisitor`. We exclude the one we override. We also had to reimplement a few extra due to `ininstance` check like the one mentioned above.